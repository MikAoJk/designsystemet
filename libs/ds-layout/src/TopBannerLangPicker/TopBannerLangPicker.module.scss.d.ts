/* automatisk genererte types */
export const list: string;
export const listItem: string;
export const srOnly: string;
