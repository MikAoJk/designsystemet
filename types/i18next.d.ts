import { resources } from '../libs/ds-core-utils/src/i18n';

declare module 'i18next' {
  interface CustomTypeOptions {
    resources: typeof resources['nb_NO'];
    keySeparator: '>';
    nsSeparator: ':';
  }
}
