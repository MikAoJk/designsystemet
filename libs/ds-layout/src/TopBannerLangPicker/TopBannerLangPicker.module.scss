@import 'libs/ds-core-designtokens/lib/designtokens/breakpoints';
@import 'libs/ds-core-utils/src/styles.scss';

.list {
  position: absolute;
  z-index: 102;
  top: 4rem;
  right: var(--spacing-m);
  left: var(--spacing-m);
  width: 100%;
  padding: 0;
  border: 1px solid var(--palette-graphite-70);
  margin-top: var(--spacing-l);
  list-style-type: none;

  @media (min-width: $breakpoint-xs) {
    right: auto;
    left: auto;
    max-width: 160px;
  }

  &::before {
    position: absolute;
    left: 50%;
    display: inline-block;
    width: 1.5rem;
    height: 1.5rem;
    border-top: 1px solid var(--palette-graphite-70);
    border-left: 1px solid var(--palette-graphite-70);
    margin-top: -14px;
    background-color: var(--palette-graphite-0);
    content: '';
    transform: rotate(45deg);

    @media (min-width: $breakpoint-xs) {
      left: var(--spacing-m);
    }
  }

  &Item {
    border-bottom: 1px solid var(--palette-graphite-70);

    &:last-of-type {
      border-bottom: none;
    }
  }
}

.srOnly {
  @include sr-only;
}
