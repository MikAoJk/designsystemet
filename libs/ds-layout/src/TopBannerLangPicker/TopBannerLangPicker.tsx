import React, { forwardRef, JSX, useEffect, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';

import { dsI18n, getCommonClassNameDefault } from '@skatteetaten/ds-core-utils';
import { MenuSVGpath } from '@skatteetaten/ds-icons';

import { ReactComponent as EnglishFlagIcon } from './Assets/en-flag.svg';
import { ReactComponent as NorwegianFlagIcon } from './Assets/no-flag.svg';
import { ReactComponent as SamiFlagIcon } from './Assets/sa-flag.svg';
import {
  getTopBannerLangPickerLocaleDefault,
  getTopBannerLangPickerShowSamiDefault,
} from './defaults';
import {
  Lang,
  TopBannerLangPickerComponent,
  TopBannerLangPickerProps,
} from './TopBannerLangPicker.types';
import { TopBannerButton } from '../TopBannerButton/TopBannerButton';
import { TopBannerLangPickerButton } from '../TopBannerLangPickerButton/TopBannerLangPickerButton';

import styles from './TopBannerLangPicker.module.scss';

export const TopBannerLangPicker = forwardRef<
  HTMLDivElement,
  TopBannerLangPickerProps
>(
  (
    {
      id,
      className = getCommonClassNameDefault(),
      lang,
      'data-testid': dataTestId,
      locale = getTopBannerLangPickerLocaleDefault(),
      showSami = getTopBannerLangPickerShowSamiDefault(),
      onLanguageClick,
    },
    ref
  ): JSX.Element => {
    const { t } = useTranslation('ds_layout', { i18n: dsI18n });

    const menuRef = useRef<HTMLUListElement>(null);
    const menuButtonRef = useRef<HTMLButtonElement>(null);

    const [selectedLang, setSelectedLang] = useState<Lang>(locale);
    const [isMenuOpen, setIsMenuOpen] = useState<boolean>(false);

    useEffect(() => {
      document.documentElement.lang = selectedLang;
    });

    useEffect(() => {
      document.documentElement.lang = selectedLang;
    }, [selectedLang]);

    const handleMenuClick = (): void => {
      setIsMenuOpen(!isMenuOpen);
    };

    const handleLanguageClick = (
      e: React.MouseEvent<HTMLButtonElement>
    ): void => {
      const lang: Lang = e.currentTarget.lang as Lang;
      setSelectedLang(lang);
      setIsMenuOpen(false);
      menuButtonRef.current?.focus();

      onLanguageClick?.(e);
    };

    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    const defaultLanguages: { [index: string]: any } = {
      nb: {
        lang: 'nb',
        displayName: 'Bokmål',
        flagIcon: <NorwegianFlagIcon />,
      },
      nn: {
        lang: 'nn',
        displayName: 'Nynorsk',
        flagIcon: <NorwegianFlagIcon />,
      },
      en: { lang: 'en', displayName: 'English', flagIcon: <EnglishFlagIcon /> },
    };

    if (showSami) {
      Object.assign(defaultLanguages, {
        sa: {
          lang: 'se',
          displayName: 'Sámegiella',
          flagIcon: <SamiFlagIcon />,
        },
      });
    }

    return (
      <div
        ref={ref}
        id={id}
        className={className}
        lang={lang}
        data-testid={dataTestId}
      >
        {/** TODO - hardkodet icon må endres til riktig flag når nytt språk velges */}
        <TopBannerButton
          ref={menuButtonRef}
          lang={selectedLang}
          svgPath={MenuSVGpath}
          ariaExpanded={isMenuOpen}
          hasArrow
          onClick={handleMenuClick}
        >
          {defaultLanguages[selectedLang].displayName}
          <span className={styles.srOnly}>{t('topbannerbutton.Menu')}</span>
        </TopBannerButton>
        {isMenuOpen && (
          <ul ref={menuRef} className={styles.list}>
            {Object.values(defaultLanguages).map((language, index) => {
              return (
                <li
                  key={`${language.lang}-${index}`}
                  className={styles.listItem}
                >
                  <TopBannerLangPicker.Button
                    lang={language.lang}
                    ariaCurrent={language.lang === selectedLang}
                    flagIcon={language.flagIcon}
                    onClick={handleLanguageClick}
                  >
                    {language.displayName}
                  </TopBannerLangPicker.Button>
                </li>
              );
            })}
          </ul>
        )}
      </div>
    );
  }
) as TopBannerLangPickerComponent;

TopBannerLangPicker.displayName = 'TopBannerLangPicker';
TopBannerLangPicker.Button = TopBannerLangPickerButton;
