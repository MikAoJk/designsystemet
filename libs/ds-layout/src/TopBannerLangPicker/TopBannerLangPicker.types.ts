import {
  ForwardRefExoticComponent,
  MouseEventHandler,
  RefAttributes,
} from 'react';

import { BaseProps, Languages } from '@skatteetaten/ds-core-utils';

import { TopBannerLangPickerButtonProps } from '../TopBannerLangPickerButton/TopBannerLangPickerButton.types';

// export const langArr = ['nb', 'nn', 'en', 'se'] as const;
// export type Lang = (typeof langArr)[number];

export interface TopBannerLangPickerProps extends BaseProps {
  /** Hvilket språk som skal være forhåndsvalgt */
  locale?: Languages;
  /** Callback når et språk trykkes på i språkvelgeren */
  onLanguageClick?: MouseEventHandler<HTMLButtonElement>;
  /** Om samisk språk skal vises eller ikke. */
  showSami?: boolean;
}

export interface TopBannerLangPickerComponent
  extends ForwardRefExoticComponent<
    TopBannerLangPickerProps & RefAttributes<HTMLDivElement>
  > {
  Button: ForwardRefExoticComponent<
    TopBannerLangPickerButtonProps & RefAttributes<HTMLButtonElement>
  >;
}
