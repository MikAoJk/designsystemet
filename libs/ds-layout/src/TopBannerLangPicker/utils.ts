import { ReactElement } from 'react';

import { Languages } from '@skatteetaten/ds-core-utils';

export const convertLocaleToLang = (locale: Languages): string => {
  return locale.substring(0, locale.indexOf('_'));
};

export type LanguageItem = {
  lang: string;
  displayName: string;
  flagIcon: ReactElement;
};

export type LanguageItems = {
  [key: string]: LanguageItem;
};
