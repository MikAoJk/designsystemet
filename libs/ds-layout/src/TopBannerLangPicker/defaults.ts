import { Lang } from './TopBannerLangPicker.types';

export const getTopBannerLangPickerShowSamiDefault = (): boolean => true;
export const getTopBannerLangPickerLocaleDefault = (): Lang => 'nb';
