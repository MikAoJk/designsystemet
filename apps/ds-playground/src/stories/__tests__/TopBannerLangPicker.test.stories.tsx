import { expect } from '@storybook/jest';
import { Meta, StoryObj } from '@storybook/react';
import { within } from '@storybook/testing-library';

// import { wrapper } from './testUtils/storybook.testing.utils';
// eslint-disable-next-line @nx/enforce-module-boundaries
import { TopBannerLangPicker } from '../../../../../libs/ds-layout/src/TopBannerLangPicker/TopBannerLangPicker';
// eslint-disable-next-line @nx/enforce-module-boundaries
import { TopBannerLangPickerProps } from '../../../../../libs/ds-layout/src/TopBannerLangPicker/TopBannerLangPicker.types';

const meta = {
  component: TopBannerLangPicker,
  title: 'Tester/TopBanner/TopBannerLangPicker (intern)',
  argTypes: {
    // Baseprops
    key: { table: { disable: true } },
    ref: { table: { disable: true } },
    className: { table: { disable: true } },
    classNames: { table: { disable: true } },
    id: { table: { disable: true } },
    lang: { table: { disable: true } },
    'data-testid': { table: { disable: true } },
    // Props
    locale: { table: { disable: true } },
    showSami: { table: { disable: true } },
    // Events
    onLanguageClick: { table: { disable: true } },
  },
} as Meta<typeof TopBannerLangPicker>;
export default meta;
type Story = StoryObj<typeof meta>;

const defaultArgs: TopBannerLangPickerProps = {};

export const WithRef = {
  name: 'With Ref (FA1)',
  args: {
    ...defaultArgs,
    ref: (instance: HTMLDivElement | null): void => {
      if (instance) {
        instance.id = 'dummyIdForwardedFromRef';
      }
    },
  },
  argTypes: {
    ref: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const container = canvas.getAllByRole('generic')[1];
    await expect(container).toHaveAttribute('id', 'dummyIdForwardedFromRef');
  },
} satisfies Story;

export const WithAttributes = {
  name: 'With Attributes (FA2-5)',
  args: {
    ...defaultArgs,
    id: 'htmlid',
    className: 'dummyClassname',
    lang: 'nb',
    'data-testid': '123ID',
  },
  argTypes: {
    id: { table: { disable: false } },
    className: { table: { disable: false } },
    lang: { table: { disable: false } },
    'data-testid': { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const container = canvas.getAllByRole('generic')[1];
    await expect(container).toHaveAttribute('id', 'htmlid');
    await expect(container).toHaveClass('dummyClassname');
    await expect(container).toHaveAttribute('lang', 'nb');
    await expect(container).toHaveAttribute('data-testid', '123ID');
  },
} satisfies Story;

export const Defaults = {
  name: 'Defaults (LanguagePicker A1, A2 delvis)',
  args: {
    ...defaultArgs,
  },
  argTypes: {},
} satisfies Story;
