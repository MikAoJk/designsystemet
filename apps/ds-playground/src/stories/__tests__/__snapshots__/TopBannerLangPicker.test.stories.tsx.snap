// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tester/TopBanner/TopBannerLangPicker (intern) Defaults smoke-test 1`] = `
<div data-test-block="true">
  <div class>
    <button class="TopBannerButton_button__5RUa7"
            lang="nb"
            aria-expanded="false"
            type="button"
    >
      <span class="TopBannerButton_iconWrapper__bu153">
        <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_icon__FyVd5"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="M3 6h18v2H3V6Zm0 5h18v2H3v-2Zm0 5h18v2H3v-2Z">
          </path>
        </svg>
        <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_arrowMobile__Z2VQa"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="m7 10 5 5 5-5H7Z">
          </path>
        </svg>
      </span>
      <span class="TopBannerButton_buttonText__guqHQ ">
        Bokmål
        <span class="TopBannerLangPicker_srOnly__33VA7">
          Meny
        </span>
      </span>
      <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_arrowDesktop__1124g"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="m7 10 5 5 5-5H7Z">
        </path>
      </svg>
    </button>
  </div>
</div>
`;

exports[`Tester/TopBanner/TopBannerLangPicker (intern) WithAttributes play-test 1`] = `
<div data-test-block="true">
  <div id="htmlid"
       class="dummyClassname"
       lang="nb"
       data-testid="123ID"
  >
    <button class="TopBannerButton_button__5RUa7"
            lang="nb"
            aria-expanded="false"
            type="button"
    >
      <span class="TopBannerButton_iconWrapper__bu153">
        <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_icon__FyVd5"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="M3 6h18v2H3V6Zm0 5h18v2H3v-2Zm0 5h18v2H3v-2Z">
          </path>
        </svg>
        <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_arrowMobile__Z2VQa"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="m7 10 5 5 5-5H7Z">
          </path>
        </svg>
      </span>
      <span class="TopBannerButton_buttonText__guqHQ ">
        Bokmål
        <span class="TopBannerLangPicker_srOnly__33VA7">
          Meny
        </span>
      </span>
      <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_arrowDesktop__1124g"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="m7 10 5 5 5-5H7Z">
        </path>
      </svg>
    </button>
  </div>
</div>
`;

exports[`Tester/TopBanner/TopBannerLangPicker (intern) WithRef play-test 1`] = `
<div data-test-block="true">
  <div class
       id="dummyIdForwardedFromRef"
  >
    <button class="TopBannerButton_button__5RUa7"
            lang="nb"
            aria-expanded="false"
            type="button"
    >
      <span class="TopBannerButton_iconWrapper__bu153">
        <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_icon__FyVd5"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="M3 6h18v2H3V6Zm0 5h18v2H3v-2Zm0 5h18v2H3v-2Z">
          </path>
        </svg>
        <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_arrowMobile__Z2VQa"
             viewbox="0 0 24 24"
             focusable="false"
             aria-hidden="true"
             role="img"
        >
          <path d="m7 10 5 5 5-5H7Z">
          </path>
        </svg>
      </span>
      <span class="TopBannerButton_buttonText__guqHQ ">
        Bokmål
        <span class="TopBannerLangPicker_srOnly__33VA7">
          Meny
        </span>
      </span>
      <svg class="Icon_systemIcon__1p1Xl Icon_systemIcon_medium__VmCFz TopBannerButton_arrowDesktop__1124g"
           viewbox="0 0 24 24"
           focusable="false"
           aria-hidden="true"
           role="img"
      >
        <path d="m7 10 5 5 5-5H7Z">
        </path>
      </svg>
    </button>
  </div>
</div>
`;
